### DISCUSSION TITLE
A more DDD-centric API for creating typed bind groups.

### DISCUSSION THREAD

**Discussion by iwoplaza** [software-mansion/TypeGPU#604] [Ideas]
`layout.populate` was always an API that did not really match with the other APIs. With the recent merge of DDD, and moving away from methods on objects in favor of functions operating on data, I would propose the following:

```ts
import tgpu from 'typegpu';
import * as d from 'typegpu/data';

const layout = tgpu.bindGroupLayout({
  a: { uniform: d.f32 },
  b: { uniform: d.u32 },
});

const root = await tgpu.init();

const aBuffer = root.createBuffer(d.f32, 0);
const bBuffer = root.createBuffer(d.u32, 0);
const bindGroup = root.createBindGroup(layout, {
  a: aBuffer,
  b: bBuffer,
});
```

### Advantages
- Matches more with the rest of the APIs (createBuffer, createTexture, ...).
- Requires the root to be present, as resources passed into the bind group are created either with a root, or a device.
- Is a function that operates on the layout, instead of being a method on the layout.

This change will allow TypeGPU to import bind group layouts defined in WESL, and use them just as well as native ones.

### Adoption
- Introduce the `root.createBindGroup` API.
- Deprecate `.populate`

**Comment by reczkok** [comment]
I like this idea. I think this is better than populate in many ways. I don't understand the need to have it on root instead of tgpu like the layout creation? I don't think it's needed at the time of creation and could introduce confusion (due to the fact that layouts are made via tgpu and bind groups are not).

**Reply by iwoplaza** [reply]
I see. To me, layouts play a role for bind groups like schemas do for buffers. One is more descriptive and corresponds more to shader functionality, and the other is a realization of that descriptor.

- Schemas -> Buffers
- Vertex Layouts -> Buffers
- Bind Group Layouts -> Bind Groups
- Slots -> Values

**Reply by iwoplaza** [reply]
So even though bind groups do not require a root to create their descriptor, the resources we pass to it do, meaning it does not belong to the strictly "descriptive" category of resources (tgpu.).

**Reply by iwoplaza** [reply]
```ts
// Descriptive Phase
//

const schema = d.struct({ ... });

const layout = tgpu.bindGroupLayout({ ... });

// Runtime Phase
//

const root = await tgpu.init();

const buffer = root.createBuffer(...);

const bindGroup = tgpu.createBindGroup(layout, ...);
// weird using tgpu. here, as we are in the runtime phase, and we need to create
// buffers before we create a bind group.

```

**Reply by reczkok** [reply]
I see. Although I don't entirely agree with the schema -> buffer analogy as schemas don't describe a WebGPU resource (and layouts have a GPUBindGroupLayout object). I agree with the phases argument. It could actually be confusing in the exact opposite direction than I originally thought. Consider me convinced 👍

**Comment by mhawryluk** [comment]
I think that's a good idea! 💡

**Comment by mhawryluk** [comment]
✅
